README.md
ed25519 HD Key
=====

Key Derivation for `ed25519`
------------
Ported from [ed25519-hd-key](https://github.com/alepop/ed25519-hd-key) js library

[SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md) - Specification

Usage
-----

```dart
import "package:ed25519_hd_key/ed25519_hd_key.dart";
import 'package:convert/convert.dart';

void main() async {
  String hexSeed =
      'fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542';
  List<int> seedBytes = hex.decode(hexSeed);
  KeyData master = await ED25519_HD_KEY.getMasterKeyFromSeed(seedBytes);
  print(hex.encode(master.key)); // 171cb88b1b3c1db25add599712e36245d75bc65a1a5c9e18d76f9f2b1eab4
  print(hex.encode(master.chainCode)); // ef70a74db9c3a5af931b5fe73ed8e1a53464133654fd55e7a66f8570b8e33c3b
  KeyData data = await ED25519_HD_KEY.derivePath("m/0'/2147483647'", seedBytes);
  print(hex.encode(data.key)); // ea4f5bfe8694d8bb74b7b59404632fd5968b774ed545e810de9c32a4fb4192f4
  print(hex.encode(data.chainCode)); // 138f0b2551bcafeca6ff2aa88ba8ed0ed8de070841f0c4ef0165df8181eaad7f
  var pb = await ED25519_HD_KEY.getPublicKey(data.key);
  print(hex.encode(pb)); // 005ba3b9ac6e90e83effcd25ac4e58a1365a9e35a3d3ae5eb07b9e4d90bcf7506d
}
```

References
----------
[SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md)

[BIP-0032](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)

[BIP-0044](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki)


CHANGELOG.md
# Changelog

## 2.2.1 (2022-08-18)

- Update dependencies

## 2.2.0 (2021-09-16)

- Web support

## 2.1.0 (2021-05-06)

- Added master secret as optional parameter

## 2.0.0 (2021-03-27)

- migrate to null safety

## 1.1.0 (2021-02-14)

- fix typo into getPublicKey method
- get rid of hex library. all public methods now accepts only Uint8List data
- migrate to cryptography library

## 1.0.1 (2019-01-24)

- add CHANGELOG
- add example

## 1.0.0 (2019-01-24)

- initial version


test/ed25519_hd_key_test.dart
import 'dart:io';
import 'dart:convert';
import "package:test/test.dart";
import '../lib/ed25519_hd_key.dart';
import 'package:convert/convert.dart';

void main() {
  Map<String, dynamic> vectors = json.decode(
      File('./test/test_vectors.json').readAsStringSync(encoding: utf8));
  final seeds = vectors.keys;

  group("Test vectors for ${seeds.first} seed", () {
    test("should have valid key and chainCode", () async {
      var master =
          await ED25519_HD_KEY.getMasterKeyFromSeed(hex.decode(seeds.first));
      expect(
          hex.encode(master.key),
          equals(
              "2b4be7f19ee27bbf30c667b642d5f4aa69fd169872f8fc3059c08ebae2eb19e7"));
      expect(
          hex.encode(master.chainCode),
          equals(
              "90046a93de5380a72b5e45010748567d5ea02bbf6522f979e05c0d8d8ca9fffb"));
    });
    for (var el in vectors[seeds.first]) {
      test("should calculate valid data for '${el['path']}' path", () async {
        KeyData data = await ED25519_HD_KEY.derivePath(
            el['path'], hex.decode(seeds.first));
        var pb = await ED25519_HD_KEY.getPublicKey(data.key);
        expect({
          "path": el['path'],
          "chainCode": hex.encode(data.chainCode),
          "key": hex.encode(data.key),
          "publicKey": hex.encode(pb),
        }, equals(el));
      });
    }
  });

  group("Test vectors for ${seeds.last} seed", () {
    test("should have valid key and chainCode", () async {
      var master =
          await ED25519_HD_KEY.getMasterKeyFromSeed(hex.decode(seeds.last));
      expect(
          hex.encode(master.key),
          equals(
              "171cb88b1b3c1db25add599712e36245d75bc65a1a5c9e18d76f9f2b1eab4012"));
      expect(
          hex.encode(master.chainCode),
          equals(
              "ef70a74db9c3a5af931b5fe73ed8e1a53464133654fd55e7a66f8570b8e33c3b"));
    });
    for (var el in vectors[seeds.last]) {
      test("should calculate valid data for '${el['path']}' path", () async {
        KeyData data =
            await ED25519_HD_KEY.derivePath(el['path'], hex.decode(seeds.last));
        var pb = await ED25519_HD_KEY.getPublicKey(data.key);
        expect({
          "path": el['path'],
          "chainCode": hex.encode(data.chainCode),
          "key": hex.encode(data.key),
          "publicKey": hex.encode(pb),
        }, equals(el));
      });
    }
  });

  group("Test optional master key change", () {
    const masterSecret = "Bitcoin seed";

    test("Using '$masterSecret' key", () async {
      var master = await ED25519_HD_KEY.getMasterKeyFromSeed(
          hex.decode(
              "cd7875cc62c027a41e030f484fb17afe9737d0eb904f7642fc1a921d5ef94344461418dd53376ea31983a29ec119b209b844fe70f6e6c86673ce2a414236a198"),
          masterSecret: masterSecret);

      expect(
          hex.encode(master.key),
          equals(
              "1352d9efc5c511f89ff262f913e58a2d42649d47246752790cbce6987e100bfe"));
    });
  });
}


lib/src/hd_key.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:pinenacl/ed25519.dart';

import 'constants.dart';
import 'key_data.dart';

// Supported curve
const ED25519_HD_KEY = _ED25519HD();

/// Implementation of ED25519 private key derivation from master private key
class _ED25519HD {
  static final _pathRegex = RegExp(r"^(m\/)?(\d+'?\/)*\d+'?$");

  const _ED25519HD();

  Future<KeyData> derivePath(String path, List<int> seedBytes,
      {int offset = HARDENED_OFFSET}) async {
    if (!_ED25519HD._pathRegex.hasMatch(path)) {
      throw ArgumentError(
          "Invalid derivation path. Expected BIP32 path format");
    }

    KeyData master = await getMasterKeyFromSeed(seedBytes);

    List<String> segments = path.split('/');
    segments = segments.sublist(1);

    KeyData result = master;

    for (String segment in segments) {
      int index = int.parse(segment.substring(0, segment.length - 1));
      result = await _getCKDPriv(result, index + offset);
    }

    return result;
  }

  Future<KeyData> getMasterKeyFromSeed(List<int> seedBytes,
          {String masterSecret = ED25519_CURVE}) =>
      _getKeys(seedBytes, utf8.encode(masterSecret));

  Future<List<int>> getPublicKey(List<int> privateKey,
      [bool withZeroByte = true]) async {
    final signature = await SigningKey.fromSeed(Uint8List.fromList(privateKey));
    final publicKey = await signature.publicKey;

    if (withZeroByte == true) {
      List<int> dataBytes = List.filled(33, 0);
      dataBytes[0] = 0x00;
      dataBytes.setRange(1, 33, publicKey);
      return dataBytes;
    } else {
      return publicKey;
    }
  }

  Future<KeyData> _getCKDPriv(KeyData data, int index) {
    Uint8List dataBytes = Uint8List(37);
    dataBytes[0] = 0x00;
    dataBytes.setRange(1, 33, data.key);
    dataBytes.buffer.asByteData().setUint32(33, index);
    return _getKeys(dataBytes, data.chainCode);
  }

  Future<KeyData> _getKeys(List<int> data, List<int> keyParameter) async {
    final hmac = Hmac(sha512, keyParameter);
    final mac = await hmac.convert(data);

    final I = mac.bytes;
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);

    return KeyData(key: IL, chainCode: IR);
  }
}


lib/src/base.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:tweetnacl/tweetnacl.dart' as ED25519;
import 'package:pointycastle/digests/sha512.dart';
import 'package:pointycastle/macs/hmac.dart';
import 'package:pointycastle/api.dart';
import 'package:hex/hex.dart';

class KeyData {
  List<int> key;
  List<int> chainCode;
  KeyData({this.key, this.chainCode});
}

const String ED25519_CURVE = 'ed25519 seed';
const int HARDENED_OFFSET = 0x80000000;

class _ED25519HD {
  static final _curveBytes = utf8.encode(ED25519_CURVE);
  static final _pathRegex = RegExp(r"^(m\/)?(\d+'?\/)*\d+'?$");

  const _ED25519HD();

  KeyData _getKeys(Uint8List data, Uint8List keyParameter) {
    final digest = SHA512Digest();
    final hmac = HMac(digest, 128)..init(KeyParameter(keyParameter));
    final I = hmac.process(data);
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);
    return KeyData(key: IL, chainCode: IR);
  }

  KeyData _getCKDPriv(KeyData data, int index) {
    Uint8List dataBytes = Uint8List(37);
    dataBytes[0] = 0x00;
    dataBytes.setRange(1, 33, data.key);
    dataBytes.buffer.asByteData().setUint32(33, index);
    return this._getKeys(dataBytes, data.chainCode);
  }

  KeyData getMasterKeyFromSeed(String seed) {
    final seedBytes = HEX.decode(seed);
    return this._getKeys(seedBytes, _ED25519HD._curveBytes);
  }

  Uint8List getBublickKey(Uint8List privateKey, [bool withZeroByte = true]) {
    final signature = ED25519.Signature.keyPair_fromSeed(privateKey);
    if (withZeroByte == true) {
      Uint8List dataBytes = Uint8List(33);
      dataBytes[0] = 0x00;
      dataBytes.setRange(1, 33, signature.publicKey);
      return dataBytes;
    } else {
      return signature.publicKey;
    }
  }

  KeyData derivePath(String path, String seed) {
    if (!_ED25519HD._pathRegex.hasMatch(path))
      throw ArgumentError(
          "Invalid derivation path. Expected BIP32 path format");
    KeyData master = this.getMasterKeyFromSeed(seed);
    List<String> segments = path.split('/');
    segments = segments.sublist(1);

    return segments.fold<KeyData>(master, (prevKeyData, indexStr) {
      int index = int.parse(indexStr.substring(0, indexStr.length - 1));
      return this._getCKDPriv(prevKeyData, index + HARDENED_OFFSET);
    });
  }
}

const ED25519_HD_KEY = const _ED25519HD();
